include/xtensor/xmasked_value.hpp

```diff
@@ -0,0 +1,817 @@
+    }
+
+    template <class T1, class B1, class T2>
+    inline auto operator==(const xmasked_value<T1, B1>& lhs, const T2& rhs) noexcept -> disable_xoptional_like<T2, bool>
+    {
+        return lhs.equal(rhs);
+    }
+
+    template <class T1, class B1, class T2, class B2>
+    inline bool operator!=(const xmasked_value<T1, B1>& lhs, const xmasked_value<T2, B2>& rhs) noexcept
+    {
+        return !lhs.equal(rhs);
+    }
+
+    template <class T1, class B1, class T2, class B2>
+    inline bool operator!=(xmasked_value<T1, B1> lhs, xtl::xoptional<T2, B2> rhs) noexcept
+    {
+        return !lhs.equal(rhs);
+    }
+
+    template <class T1, class B1, class T2, class B2>
+    inline bool operator!=(xtl::xoptional<T1, B1> lhs, xmasked_value<T2, B2> rhs) noexcept
+    {
+        return !rhs.equal(lhs);
+    }
+
+    template <class T1, class T2, class B2>
+    inline auto operator!=(const T1& lhs, const xmasked_value<T2, B2>& rhs) noexcept -> disable_xoptional_like<T1, bool>
+    {
+        return !rhs.equal(lhs);
+    }
+
+    template <class T1, class B1, class T2>
+    inline auto operator!=(const xmasked_value<T1, B1>& lhs, const T2& rhs) noexcept -> disable_xoptional_like<T2, bool>
+    {
+        return !lhs.equal(rhs);
+    }
+
+    template <class T, class B>
+    inline auto operator+(const xmasked_value<T, B>& e) noexcept
+        -> xmasked_value<std::decay_t<T>, std::decay_t<B>>
+    {
+        return xmasked_value<std::decay_t<T>, std::decay_t<B>>(e.value(), e.has_value());
+    }
+
+    template <class T, class B>
+    inline auto operator-(const xmasked_value<T, B>& e) noexcept
+        -> xmasked_value<std::decay_t<T>, std::decay_t<B>>
+    {
+        return xmasked_value<std::decay_t<T>, std::decay_t<B>>(-e.value(), e.has_value());
+    }
+
+    template <class T, class B>
+    inline auto operator~(const xmasked_value<T, B>& e) noexcept
+        -> xmasked_value<std::decay_t<T>>
+    {
+        using value_type = std::decay_t<T>;
+        return e.has_value() ? masked_value(~e.value()) : missing<value_type>();
+    }
+
+    template <class T, class B>
+    inline auto operator!(const xmasked_value<T, B>& e) noexcept
+        -> xmasked_value<bool>
+    {
+        return e.has_value() ? masked_value(!e.value()) : missing<bool>();
+    }
+
+    template <class T, class B, class OC, class OT>
+    inline std::basic_ostream<OC, OT>& operator<<(std::basic_ostream<OC, OT>& out, xmasked_value<T, B> v)
+    {
+        if (v.has_value())
+        {
+            out << v.value();
+        }
+        else
+        {
+            out << "N/A";
+        }
+        return out;
+    }
+
+#define DEFINE_OPERATOR(OP)                                                                                      \
+    template <class T1, class B1, class T2, class B2>                                                            \
+    inline auto operator OP(const xmasked_value<T1, B1>& e1, const xtl::xoptional<T2, B2>& e2) noexcept          \
```

Comment:
can be removed

Reply:
But this is needed if we want to be able to add an `xoptional_assembly` to an `xmasked_view`, right?

Reply:
This should not, the `xoptional` would be used as the `value_type` of the `xmasked_value`.

Reply:
Note that this required using `promote_type` instead of `std::common_type` when computing the common type of the `xmasked_view::value_type` and another type.

Reply:
Nop, the `value_type` is always the `inner_type` of `xoptional` in `xmasked_view`.

- If you make a `xmasked_view` on an `xoptional_assembly<xarray<double>, xarray<bool>>` then the `reference_type` of `xmasked_view` is `xmasked_value<double&>` and the `value_type` of `xmasked_value` is `double`
- If you make a `xmasked_view` on an `xarray<double>` then the `reference_type` of `xmasked_view` is again `xmasked_value<double&>` and the `value_type` of `xmasked_value` is `double`

Reply:
I don't want the `value_type` of `xmasked_value` to be `xoptional`. 
In xframe, if I make a `where_view` of my variable, if want the underlying data to keep the "same" underlying types, at least the same API. The underlying types of my variable is `xoptional<double&, bool&>`, and the underlying type of the `where_view` of my variable is `xmasked_value<double&, bool&>`

Reply:
The behavior of `xmasked_view` should be independent from its underlying `value_type`. `xmasked_value` should not handle arithmetic with `xoptional` as a special case, everything should be homogeneous. If I have a `xmasked_view` on a expression whose `value_type` is T, then the `value_type` of the `xmasked_view` must be T, whatever T is.

More generally, we want the `xmasked_value` to be "on top" of the `value_type`, not at the same level, to avoid confusion with potential `xoptional` value type, and this is true for `xframe` too.

The usage of the mask of `xmasked_view` as an overload of the mask of `xoptional` is now done in the conversion operator, so no need to handle `xoptional` as a specific case anywhere else.

Reply:
Ok, let's do that. But are you sure that, if we do it like you say, when I will add an `xmasked_view` containing `xmasked_value`s of `xoptional`s to an `xoptional_assembly` it will work perfectly just using the conversion from `xmasked_value` to `xoptional`?

Reply:
It should, if not, lets modify the `xmasked_view` so it has a correct behavior ;)